stdlib/runtime/exception.gr

```diff
@@ -0,0 +1,103 @@
+/* grainc-flags --compilation-mode=runtime */
+
+import WasmI32, {
+  eq as (==),
+  add as (+)
+} from "runtime/unsafe/wasmi32"
+
+import foreign wasm fd_write: (WasmI32, WasmI32, WasmI32, WasmI32) -> WasmI32 from "wasi_snapshot_preview1"
+
+enum Option<a> { Some(a), None }
+
+export let mut printers = 0n
+
+export let registerBasePrinter = (f) => {
+  let mut current = printers
+  while (true) {
+    // There will be at least one printer registered by the time this is called
+    let (_, next) = WasmI32.toGrain(current) : (Exception -> Option<String>, WasmI32)
+    if (next == 0n) {
+      // Using a tuple in runtime mode is typically disallowed as there is no way
+      // to reclaim the memory, but this function is only called once
+      let newBase = (WasmI32.fromGrain(f), 0n)
+      WasmI32.store(current, WasmI32.fromGrain(newBase), 12n)
+      break
+    }
+    current = next
+  }
+}
+
+export let dangerouslyRegisterPrinter = (f) => {
+  // This is dangerous because it leaks runtime memory.
+  // This should only be called by this module and Pervasives.
+  printers = WasmI32.fromGrain((f, printers))
+}
```

Comment:
Is there a way we can make these completely unavailable to consumers? Calling these a 2nd time totally screws everything up, right?

Reply:
Not as of right now, but generally people shouldn't ever be importing anything from `runtime`. We could achieve that though with a compiler flag or something similar that by default disables the ability to import from the runtime.

The only harm in this being called more than once is a memory leak. Not the worst thing in the world (but still not great).